llumnix/global_scheduler/migration_scheduler.py

```diff
@@ -43,13 +59,37 @@ def __init__(self,
 
         self.num_instances = 0
         self.instance_id_set: Set[str] = set()
+        self.instance_id_type_set: Dict[InstanceType, Set[str]] = {instance_type: set() for instance_type in InstanceType}
+        self.constraint_prefill_instance_num = constraint_prefill_instance_num
         # instance info args
         self.instance_info: Dict[str, InstanceInfo] = None
-        self.sorted_instance_infos: List[InstanceInfo] = None
+        self.sorted_instance_infos: Dict[str, List[InstanceInfo]] = {instance_type: set() for instance_type in InstanceType}
 
-    def pair_migration(self) -> List[Tuple[str, str]]:
+    def pair_migration(self, migration_target:str) -> List[Tuple[str, str]]:
         self._sort_instance_infos(descending=False)
-        return self.pair_migration_policy.pair_migration(self.sorted_instance_infos)
+        sorted_src_instance_infos, sorted_dst_instance_infos, strict_pre_migration = self._get_migration_settings(migration_target)
+        return self.pair_migration_policy.pair_migration(sorted_src_instance_infos, sorted_dst_instance_infos, strict_pre_migration)
+
+    def _get_migration_settings(self, migration_target:str) -> Dict[str, InstanceInfo]:
+        strict_pre_migration = True
+        if migration_target == PairMigrationConstraints.NO_CONSTRAINTS:
+            # migrate out instances
+            sorted_src_instance_infos = [i for i in reversed(self.sorted_instance_infos[InstanceType.NO_CONSTRAINTS])
+                                if i.num_killed_requests > 0 or i.instance_load_migrate > self.migrate_out_load_threshold]
+            # migrate in instances
+            sorted_dst_instance_infos = [i for i in self.sorted_instance_infos[InstanceType.NO_CONSTRAINTS]
+                               if i.num_killed_requests == 0 and i.instance_load_migrate < self.migrate_out_load_threshold]
+        elif migration_target == PairMigrationConstraints.PREFILL_2_DECODING:
+            sorted_src_instance_infos = list(reversed(self.sorted_instance_infos[InstanceType.PREFILL]))
+            sorted_dst_instance_infos = [i for i in self.sorted_instance_infos[InstanceType.DECODE]
+                                            if i.num_killed_requests == 0]
+            # TODO[xinyi]: Considering decoding instances load, try to decode on the prefill instance(set strict_pre_migration as False).
+        elif migration_target == PairMigrationConstraints.DECODING_2_DECODING:
+            sorted_src_instance_infos = [i for i in reversed(self.sorted_instance_infos[InstanceType.DECODE])
+                                if i.num_killed_requests > 0 or i.instance_load_migrate > self.migrate_out_load_threshold]
+            sorted_dst_instance_infos = [i for i in self.sorted_instance_infos[InstanceType.DECODE]
+                               if i.num_killed_requests == 0 and i.instance_load_migrate < self.migrate_out_load_threshold]
+        return sorted_src_instance_infos, sorted_dst_instance_infos, strict_pre_migration
```

Comment:
why we return strict_pre_migration here? It is a bit odd to return instance_infos and strict_pre_migration in a single function, because they have no strong relations.

Reply:
> why we return strict_pre_migration here? It is a bit odd to return instance_infos and strict_pre_migration in a single function, because they have no strong relations.

The function is used to provide migration settings ( I think maybe whether to force migration is also a type of migration configuration). The src and dst instance infos are selected based on load, and whether to enable strict pre-migration on src inst are also determined by the load of the dst inst (to be implemented in the next step).